File: clients/src/test/java/org/apache/kafka/common/requests/TxnOffsetCommitRequestTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kafka.common.requests;

import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.Uuid;
import org.apache.kafka.common.errors.UnsupportedVersionException;
import org.apache.kafka.common.message.TxnOffsetCommitRequestData;
import org.apache.kafka.common.message.TxnOffsetCommitRequestData.TxnOffsetCommitRequestPartition;
import org.apache.kafka.common.message.TxnOffsetCommitRequestData.TxnOffsetCommitRequestTopic;
import org.apache.kafka.common.message.TxnOffsetCommitResponseData;
import org.apache.kafka.common.protocol.ApiKeys;
import org.apache.kafka.common.protocol.Errors;
import org.apache.kafka.common.requests.TxnOffsetCommitRequest.CommittedOffset;
import org.apache.kafka.common.utils.annotation.ApiKeyVersionsSource;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class TxnOffsetCommitRequestTest extends OffsetCommitRequestTest {

    private static final Map<TopicPartition, CommittedOffset> OFFSETS = new HashMap<>();
    private static TxnOffsetCommitRequest.Builder builder;
    private static TxnOffsetCommitRequest.Builder builderWithGroupMetadata;

    @BeforeEach
    @Override
    public void setUp() {
        super.setUp();
        OFFSETS.clear();
        OFFSETS.put(new TopicPartition(topicOne, partitionOne),
            new CommittedOffset(
                offset,
                metadata,
                Optional.of((int) leaderEpoch)));
        OFFSETS.put(new TopicPartition(topicTwo, partitionTwo),
            new CommittedOffset(
                offset,
                metadata,
                Optional.of((int) leaderEpoch)));

        String transactionalId = "transactionalId";
        int producerId = 10;
        short producerEpoch = 1;
        TxnOffsetCommitRequestData data = new TxnOffsetCommitRequestData()
            .setTransactionalId(transactionalId)
            .setGroupId(groupId)
            .setProducerId(producerId)
            .setProducerEpoch(producerEpoch)
            .setTopics(TxnOffsetCommitRequest.getTopics(OFFSETS));
        builder = TxnOffsetCommitRequest.Builder.forTopicNames(data, true);

        int generationId = 5;
        TxnOffsetCommitRequestData dataWithGroupMetadata = new TxnOffsetCommitRequestData()
            .setTransactionalId(transactionalId)
            .setGroupId(groupId)
            .setProducerId(producerId)
            .setProducerEpoch(producerEpoch)
            .setMemberId(memberId)
            .setGenerationIdOrMemberEpoch(generationId)
            .setGroupInstanceId(groupInstanceId)
            .setTopics(TxnOffsetCommitRequest.getTopics(OFFSETS));
        builderWithGroupMetadata = TxnOffsetCommitRequest.Builder.forTopicNames(dataWithGroupMetadata, true);
    }

    @ParameterizedTest
    @ApiKeyVersionsSource(apiKey = ApiKeys.TXN_OFFSET_COMMIT, toVersion = 5)
    public void testConstructor(short version) {
        List<TxnOffsetCommitRequestTopic> expectedTopics = List.of(
            new TxnOffsetCommitRequestTopic()
                .setName(topicOne)
                .setPartitions(List.of(
                    new TxnOffsetCommitRequestPartition()
                        .setPartitionIndex(partitionOne)
                        .setCommittedOffset(offset)
                        .setCommittedLeaderEpoch(leaderEpoch)
                        .setCommittedMetadata(metadata))),
            new TxnOffsetCommitRequestTopic()
                .setName(topicTwo)
                .setPartitions(List.of(
                    new TxnOffsetCommitRequestPartition()
                        .setPartitionIndex(partitionTwo)
                        .setCommittedOffset(offset)
                        .setCommittedLeaderEpoch(leaderEpoch)
                        .setCommittedMetadata(metadata))));

        final TxnOffsetCommitRequest request;
        if (version < 3) {
            request = builder.build(version);
        } else {
            request = builderWithGroupMetadata.build(version);
        }
        assertEquals(OFFSETS, request.offsets());
        assertEquals(expectedTopics, TxnOffsetCommitRequest.getTopics(request.offsets()));

        var response = request.getErrorResponse(throttleTimeMs, Errors.NOT_COORDINATOR.exception());

        assertEquals(Map.of(Errors.NOT_COORDINATOR, 2), response.errorCounts());
        assertEquals(throttleTimeMs, response.throttleTimeMs());
    }

    @Test
    @Override
    public void testGetErrorResponse() {
        var topicOneId = Uuid.randomUuid();
        var topicTwoId = Uuid.randomUuid();

        var data = new TxnOffsetCommitRequestData()
            .setTransactionalId("transactionalId")
            .setGroupId(groupId)
            .setProducerId(10L)
            .setProducerEpoch((short) 1)
            .setTopics(List.of(
                new TxnOffsetCommitRequestTopic()
                    .setTopicId(topicOneId)
                    .setName(topicOne)
                    .setPartitions(List.of(
                        new TxnOffsetCommitRequestPartition()
                            .setPartitionIndex(partitionOne)
                            .setCommittedOffset(offset))),
                new TxnOffsetCommitRequestTopic()
                    .setTopicId(topicTwoId)
                    .setName(topicTwo)
                    .setPartitions(List.of(
                        new TxnOffsetCommitRequestPartition()
                            .setPartitionIndex(partitionTwo)
                            .setCommittedOffset(offset)))));

        var expectedResponseData = new TxnOffsetCommitResponseData()
            .setTopics(List.of(
                new TxnOffsetCommitResponseData.TxnOffsetCommitResponseTopic()
                    .setTopicId(topicOneId)
                    .setName(topicOne)
                    .setPartitions(List.of(
                        new TxnOffsetCommitResponseData.TxnOffsetCommitResponsePartition()
                            .setPartitionIndex(partitionOne)
                            .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code()))),
                new TxnOffsetCommitResponseData.TxnOffsetCommitResponseTopic()
                    .setTopicId(topicTwoId)
                    .setName(topicTwo)
                    .setPartitions(List.of(
                        new TxnOffsetCommitResponseData.TxnOffsetCommitResponsePartition()
                            .setPartitionIndex(partitionTwo)
                            .setErrorCode(Errors.UNKNOWN_MEMBER_ID.code())))));

        assertEquals(expectedResponseData, TxnOffsetCommitRequest.getErrorResponse(data, Errors.UNKNOWN_MEMBER_ID));

        var request = TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(data, true).build();
        var response = request.getErrorResponse(throttleTimeMs, Errors.UNKNOWN_MEMBER_ID.exception());
        assertEquals(expectedResponseData.setThrottleTimeMs(throttleTimeMs), response.data());
    }

    @ParameterizedTest
    @ApiKeyVersionsSource(apiKey = ApiKeys.TXN_OFFSET_COMMIT, toVersion = 5)
    public void testVersionSupportForGroupMetadata(short version) {
        assertDoesNotThrow(() -> builder.build(version));
        if (version >= 3) {
            assertDoesNotThrow(() -> builderWithGroupMetadata.build(version));
        } else {
            assertEquals("Broker doesn't support group metadata commit API on version " + version +
                ", minimum supported request version is 3 which requires brokers to be on version 2.5 or above.",
                assertThrows(UnsupportedVersionException.class, () -> builderWithGroupMetadata.build(version)).getMessage());
        }
    }

    @ParameterizedTest
    @ApiKeyVersionsSource(apiKey = ApiKeys.TXN_OFFSET_COMMIT)
    public void testForTopicIdsOrNamesWithTopicNameOnly(short version) {
        var data = new TxnOffsetCommitRequestData()
            .setTransactionalId("tx")
            .setGroupId(groupId)
            .setProducerId(1)
            .setProducerEpoch((short) 0)
            .setTopics(List.of(
                new TxnOffsetCommitRequestTopic()
                    .setName("foo")
                    .setPartitions(List.of(
                        new TxnOffsetCommitRequestPartition()
                            .setPartitionIndex(0)
                            .setCommittedOffset(0L)))));

        if (version >= 6) {
            assertThrows(UnsupportedVersionException.class,
                () -> TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(data, true).build(version));
        } else {
            assertDoesNotThrow(
                () -> TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(data, true).build(version));
        }
    }

    @ParameterizedTest
    @ApiKeyVersionsSource(apiKey = ApiKeys.TXN_OFFSET_COMMIT)
    public void testForTopicIdsOrNamesWithTopicIdOnly(short version) {
        var topicId = Uuid.randomUuid();
        var data = new TxnOffsetCommitRequestData()
            .setTransactionalId("tx")
            .setGroupId(groupId)
            .setProducerId(1)
            .setProducerEpoch((short) 0)
            .setTopics(List.of(
                new TxnOffsetCommitRequestTopic()
                    .setTopicId(topicId)
                    .setPartitions(List.of(
                        new TxnOffsetCommitRequestPartition()
                            .setPartitionIndex(0)
                            .setCommittedOffset(0L)))));

        if (version >= 6) {
            var request = TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(data, true).build(version);
            assertEquals(data, request.data());
        } else {
            assertThrows(UnsupportedVersionException.class,
                () -> TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(data, true).build(version));
        }
    }

    @Test
    public void testForTopicNamesCapsAtTransactionV1WhenTransactionV2IsDisabled() {
        var builder = TxnOffsetCommitRequest.Builder.forTopicNames(
            new TxnOffsetCommitRequestData(),
            false
        );
        assertEquals(TxnOffsetCommitRequest.LAST_STABLE_VERSION_BEFORE_TRANSACTION_V2, builder.latestAllowedVersion());
    }

    @Test
    public void testForTopicNamesCapsAtV5WhenTransactionV2IsEnabled() {
        var builder = TxnOffsetCommitRequest.Builder.forTopicNames(
            new TxnOffsetCommitRequestData(),
            true
        );
        assertEquals((short) 5, builder.latestAllowedVersion());
    }

    @Test
    public void testForTopicIdsOrNamesCapsAtTransactionV1WhenTransactionV2IsDisabled() {
        var builder = TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(
            new TxnOffsetCommitRequestData(),
            false
        );
        assertEquals(TxnOffsetCommitRequest.LAST_STABLE_VERSION_BEFORE_TRANSACTION_V2, builder.latestAllowedVersion());
    }

    @Test
    public void testForTopicIdsOrNamesUsesLatestVersionWhenTransactionV2IsEnabled() {
        var builder = TxnOffsetCommitRequest.Builder.forTopicIdsOrNames(
            new TxnOffsetCommitRequestData(),
            true
        );
        assertEquals(ApiKeys.TXN_OFFSET_COMMIT.latestVersion(), builder.latestAllowedVersion());
    }
}

File: core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package kafka.server

import kafka.utils.TestUtils
import org.apache.kafka.common.errors.UnsupportedVersionException
import org.apache.kafka.common.protocol.{ApiKeys, Errors}
import org.apache.kafka.common.requests.{EndTxnRequest, JoinGroupRequest}
import org.apache.kafka.common.test.ClusterInstance
import org.apache.kafka.common.test.api.{ClusterConfigProperty, ClusterTest, ClusterTestDefaults, Type}
import org.apache.kafka.common.utils.ProducerIdAndEpoch
import org.apache.kafka.coordinator.group.GroupCoordinatorConfig
import org.apache.kafka.coordinator.transaction.TransactionLogConfig
import org.junit.jupiter.api.Assertions.{assertNotEquals, assertThrows}

@ClusterTestDefaults(
  types = Array(Type.KRAFT),
  serverProperties = Array(
    new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, value = "1"),
    new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
    new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_PARTITIONS_CONFIG, value = "1"),
    new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
  )
)
class TxnOffsetCommitRequestTest(cluster:ClusterInstance) extends GroupCoordinatorBaseRequestTest(cluster) {

  @ClusterTest
  def testTxnOffsetCommitWithNewConsumerGroupProtocol(): Unit = {
    testTxnOffsetCommit(true)
  }

  @ClusterTest
  def testTxnOffsetCommitWithOldConsumerGroupProtocol(): Unit = {
    testTxnOffsetCommit(false)
  }

  @ClusterTest
  def testDelayedTxnOffsetCommitWithBumpedEpochIsRejectedWithNewConsumerGroupProtocol(): Unit = {
    testDelayedTxnOffsetCommitWithBumpedEpochIsRejected(true)
  }

  @ClusterTest
  def testDelayedTxnOffsetCommitWithBumpedEpochIsRejectedWithOldConsumerGroupProtocol(): Unit = {
    testDelayedTxnOffsetCommitWithBumpedEpochIsRejected(false)
  }

  private def testTxnOffsetCommit(useNewProtocol: Boolean): Unit = {
    val topic = "topic"
    val partition = 0
    val transactionalId = "txn"
    val groupId = "group"

    // Creates the __consumer_offsets and __transaction_state topics because it won't be created automatically
    // in this test because it does not use FindCoordinator API.
    createOffsetsTopic()
    createTransactionStateTopic()

    // Join the consumer group. Note that we don't heartbeat here so we must use
    // a session long enough for the duration of the test.
    val (memberId: String, memberEpoch: Int) = joinConsumerGroup(groupId, useNewProtocol)
    assertNotEquals(JoinGroupRequest.UNKNOWN_MEMBER_ID, memberId)
    assertNotEquals(JoinGroupRequest.UNKNOWN_GENERATION_ID, memberEpoch)

    createTopic(topic, 1)

    for (version <- 0 to ApiKeys.TXN_OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
      // Verify that the TXN_OFFSET_COMMIT request is processed correctly when member id is UNKNOWN_MEMBER_ID
      // and generation id is UNKNOWN_GENERATION_ID under all api versions.
      verifyTxnCommitAndFetch(
        topic = topic,
        partition = partition,
        transactionalId = transactionalId,
        groupId = groupId,
        memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID,
        generationId = JoinGroupRequest.UNKNOWN_GENERATION_ID,
        offset = 100 + version,
        version = version.toShort,
        expectedTxnCommitError = Errors.NONE
      )

      if (version >= 3) {
        // Verify that the TXN_OFFSET_COMMIT request is processed correctly when the member ID
        // and generation ID are known. This validation starts from version 3, as the member ID
        // must not be empty from version 3 onwards.
        verifyTxnCommitAndFetch(
          topic = topic,
          partition = partition,
          transactionalId = transactionalId,
          groupId = groupId,
          memberId = memberId,
          generationId = memberEpoch,
          offset = 200 + version,
          version = version.toShort,
          expectedTxnCommitError = Errors.NONE
        )

        // Verify TXN_OFFSET_COMMIT request failed with incorrect memberId.
        verifyTxnCommitAndFetch(
          topic = topic,
          partition = partition,
          transactionalId = transactionalId,
          groupId = groupId,
          memberId = "non-exist",
          generationId = memberEpoch,
          offset = 200 + version,
          version = version.toShort,
          expectedTxnCommitError = Errors.UNKNOWN_MEMBER_ID
        )

        // Verify TXN_OFFSET_COMMIT request failed with incorrect generationId.
        verifyTxnCommitAndFetch(
          topic = topic,
          partition = partition,
          transactionalId = transactionalId,
          groupId = groupId,
          memberId = memberId,
          generationId = 100,
          offset = 200 + version,
          version = version.toShort,
          expectedTxnCommitError = Errors.ILLEGAL_GENERATION
        )
      } else {
        // Verify that the TXN_OFFSET_COMMIT request failed when group metadata is set under version 3.
        assertThrows(classOf[UnsupportedVersionException], () =>
          verifyTxnCommitAndFetch(
            topic = topic,
            partition = partition,
            transactionalId = transactionalId,
            groupId = groupId,
            memberId = memberId,
            generationId = memberEpoch,
            offset = 200 + version,
            version = version.toShort,
            expectedTxnCommitError = Errors.NONE
          )
        )
      }
    }
  }

  private def verifyTxnCommitAndFetch(
    topic: String,
    partition: Int,
    transactionalId: String,
    groupId: String,
    memberId: String,
    generationId: Int,
    offset: Long,
    version: Short,
    expectedTxnCommitError: Errors
  ): Unit = {
    var producerIdAndEpoch: ProducerIdAndEpoch = null
    // Wait until the coordinator finishes loading.
    TestUtils.waitUntilTrue(() =>
      try {
        producerIdAndEpoch = initProducerId(
          transactionalId = transactionalId,
          producerIdAndEpoch = ProducerIdAndEpoch.NONE,
          expectedError = Errors.NONE
        )
        true
      } catch {
        case _: Throwable => false
      }, "initProducerId request failed"
    )

    addOffsetsToTxn(
      groupId = groupId,
      producerId = producerIdAndEpoch.producerId,
      producerEpoch = producerIdAndEpoch.epoch,
      transactionalId = transactionalId
    )

    val originalOffset = fetchOffset(groupId, topic, partition)

    commitTxnOffset(
      groupId = groupId,
      memberId = memberId,
      generationId = generationId,
      producerId = producerIdAndEpoch.producerId,
      producerEpoch = producerIdAndEpoch.epoch,
      transactionalId = transactionalId,
      topic = topic,
      partition = partition,
      offset = offset,
      expectedError = expectedTxnCommitError,
      version = version
    )

    endTxn(
      producerId = producerIdAndEpoch.producerId,
      producerEpoch = producerIdAndEpoch.epoch,
      transactionalId = transactionalId,
      isTransactionV2Enabled = false,
      committed = true,
      expectedError = Errors.NONE
    )

    val expectedOffset = if (expectedTxnCommitError == Errors.NONE) offset else originalOffset

    TestUtils.waitUntilTrue(() =>
      try {
        fetchOffset(groupId, topic, partition) == expectedOffset
      } catch {
        case _: Throwable => false
      }, "txn commit offset validation failed"
    )
  }

  private def testDelayedTxnOffsetCommitWithBumpedEpochIsRejected(useNewProtocol: Boolean): Unit = {
    val topic = "topic"
    val partition = 0
    val transactionalId = "txn"
    val groupId = "group"
    val offset = 100L

    // Creates the __consumer_offsets and __transaction_state topics because it won't be created automatically
    // in this test because it does not use FindCoordinator API.
    createOffsetsTopic()
    createTransactionStateTopic()

    // Join the consumer group. Note that we don't heartbeat here so we must use
    // a session long enough for the duration of the test.
    val (memberId: String, memberEpoch: Int) = joinConsumerGroup(groupId, useNewProtocol)
    assertNotEquals(JoinGroupRequest.UNKNOWN_MEMBER_ID, memberId)
    assertNotEquals(JoinGroupRequest.UNKNOWN_GENERATION_ID, memberEpoch)

    createTopic(topic, 1)

    for (version <- ApiKeys.TXN_OFFSET_COMMIT.oldestVersion to ApiKeys.TXN_OFFSET_COMMIT.latestVersion(isUnstableApiEnabled)) {
      val useTV2 = version > EndTxnRequest.LAST_STABLE_VERSION_BEFORE_TRANSACTION_V2

      // Initialize producer. Wait until the coordinator finishes loading.
      var producerIdAndEpoch: ProducerIdAndEpoch = null
      TestUtils.waitUntilTrue(() =>
        try {
          producerIdAndEpoch = initProducerId(
            transactionalId = transactionalId,
            producerIdAndEpoch = ProducerIdAndEpoch.NONE,
            expectedError = Errors.NONE
          )
          true
        } catch {
          case _: Throwable => false
        }, "initProducerId request failed"
      )

      addOffsetsToTxn(
        groupId = groupId,
        producerId = producerIdAndEpoch.producerId,
        producerEpoch = producerIdAndEpoch.epoch,
        transactionalId = transactionalId
      )

      // Complete the transaction.
      endTxn(
        producerId = producerIdAndEpoch.producerId,
        producerEpoch = producerIdAndEpoch.epoch,
        transactionalId = transactionalId,
        isTransactionV2Enabled = useTV2,
        committed = true,
        expectedError = Errors.NONE
      )

      // Start a new transaction. Wait for the previous transaction to complete.
      TestUtils.waitUntilTrue(() =>
        try {
          addOffsetsToTxn(
            groupId = groupId,
            producerId = producerIdAndEpoch.producerId,
            producerEpoch = if (useTV2) (producerIdAndEpoch.epoch + 1).toShort else producerIdAndEpoch.epoch,
            transactionalId = transactionalId
          )
          true
        } catch {
          case _: Throwable => false
        }, "addOffsetsToTxn request failed"
      )

      // Committing offset with old epoch succeeds for TV1 and fails for TV2.
      commitTxnOffset(
        groupId = groupId,
        memberId = if (version >= 3) memberId else JoinGroupRequest.UNKNOWN_MEMBER_ID,
        generationId = if (version >= 3) 1 else JoinGroupRequest.UNKNOWN_GENERATION_ID,
        producerId = producerIdAndEpoch.producerId,
        producerEpoch = producerIdAndEpoch.epoch,
        transactionalId = transactionalId,
        topic = topic,
        partition = partition,
        offset = offset,
        expectedError = if (useTV2) Errors.INVALID_PRODUCER_EPOCH else Errors.NONE,
        version = version.toShort
      )

      // Complete the transaction.
      endTxn(
        producerId = producerIdAndEpoch.producerId,
        producerEpoch = if (useTV2) (producerIdAndEpoch.epoch + 1).toShort else producerIdAndEpoch.epoch,
        transactionalId = transactionalId,
        isTransactionV2Enabled = useTV2,
        committed = true,
        expectedError = Errors.NONE
      )
    }
  }
}
